### PR TITLE
[Google Blockly] Create SVG frames for modal editor workspaces

### PR DESCRIPTION
**Pull request by mikeharv**
## Background

This is part of the work to complete the new modal function editor for Google Blockly labs. Because the UI is significantly different, leaning on block fields and flyouts instead of standard text fields, we needed a new design.

We already have a `blockSvgFrame` class that adds a frame around both procedure definitions (via the base class class) and unused blocks (via the extension `blockSvgUnused`):
image
image

Students are already familiar with these frames as a way to encapsulate functions. Examples from our curricula:
https://studio.code.org/s/hero/lessons/1/levels/8?no_autoplay=true
https://studio.code.org/s/artist/lessons/1/levels/8
https://studio.code.org/s/csc-function-machines-pilot/lessons/3/levels/6/sublevel/1

Therefore, it makes sense to consider using a similar design to represent functions in other contexts. This PR allows us to use similar frames for the workspace of the modal function editor. These frames stretch to fill the workspace, in hopes of better suggesting that your sole focus in this mode is to view or edit the procedure definition.
image

## High-Level Approach
* Move core logic from `blockSvgFrame` to a new `svgFrame` base class.
* Create a new `workspaceSvgFrame` class that also extends the new base class.

## Other Technical Details
The new svg frame uses the same styles as the  default block frame used for functions, for now.
Workspace frames are different from block frames in a few important ways:
* They are created on a workspace rather than a block.
* Rather than surrounding their parent element, workspace frames fit within it.
* The rendered height and width are based on the viewable area of the workspace, unless the block contents is larger.
* Workspace frames need to listen for certain Blockly and browser events so that they can be dynamically resized.
* Workspace frames need to be removable so that we can switch between functions and behaviors.

While doing this work, I realized we could remove the existing `render` parameters (the svg group of the host element and an RTL boolean), as these arguments were always obtainable from the `this` context of the class instance. I then added _new_ parameters for height/width. This allowed me to have all frame versions use the base `render` (with the workspace version including an override and `super.render` call with custom starting dimensions).

## Screen captures

**Function Editor**
image

**Behavior Editor**
image

**Workspace content larger than visible workspace area**

**Frame size adjusts to workspace content changes**

**Frame size adjust to theme changes**

**Frame size adjusts to browser window resizing**

While doing this work, I also found and fixed two rendering bugs related to unused blocks.
First, very small blocks were being created with frames too small for the text. Second, switching to high contrast themes and back caused improperly sized frames. The first bugfix required properly setting a minimum width based on the frame's `frameText_`'s bounding rectangle. The second bugfix required adding an event listener for theme changes.
 
_Staging:_
imageimage

_Fixed on feature branch:_
image
image

## Alternative approaches considered
I first tried modifying the existing block frames so that they simply expanded for blocks in a modal editor: 9d5aad59151439a60ee6e097693c33a2ace13465
Unfortunately, this came with an unintended side effect. If other disconnected blocks were placed around the procedure code on the modal editor workspace, they would disappear from view if the user clicked the definition block or its children. This was happening because Blockly brings the selected block forward. The opaque frame was also brought forward, and because it filled the whole view area, it made it impossible to see or select other blocks.
After this didn't work, I manually reverted many of the changes and created a copy of `blockSvgFrame` class for workspaces. This worked out well, but there was a lot of repeated logic between the classes. Finally I chose the approach above where a new base class was created as well.

## Follow-up work

* Decide if we want to update the styles of these frames, for example, headers on function frames could match the block color of the current theme. (image shows a potential change, not present on this branch)
image (161)

* Decide if we want to update the frame body color when in dark mode (image shows current colors used on this branch)
image (160)

* Confirm that the text in the header is drawn correctly after we start correctly passing in RTL options when creating the workspace. _Note:_ When opening the modal in an RTL language, we currently still show the toolbox on the left and the close/delete buttons on the right. As such, I think it makes sense to leave the "Function" title on the left of the frame. We will likely want to reverse that once we are injecting with the proper options. (Image reflects current behavior on this branch.)
image

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/functionEditor.js [hunk 1]
```diff
@@ -166,11 +170,24 @@ export default class FunctionEditor {
...
       };
+
       this.block = Blockly.serialization.blocks.append(
         this.addEditorWorkspaceBlockConfig(newDefinitionBlock),
         this.editorWorkspace
       );
     }
+    const type = procedureType || existingProcedureBlock.type;
+    const isBehavior = type === BLOCK_TYPES.behaviorDefinition;
+    const existingFrame = this.editorWorkspace.svgFrame_;
```
Comment (molly-moen) [inline apps/src/blockly/addons/functionEditor.js]: nit: move the check for and disposal of `existingFrame` to `clearEditorWorkspace`

File: apps/src/blockly/addons/workspaceSvgFrame.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  addBrowserResizeListener() {
+    window.addEventListener('resize', () => {
+      // Frame size can depend upon workspace size, so we re-render.
+      this.render(this.element_.svgGroup_, this.element_.RTL);
+    });
+  }
+
+  /**
+   * Overrides the standard render to create a frame within the element, rather than around it.
```
Comment (molly-moen) [inline apps/src/blockly/addons/workspaceSvgFrame.js:46]: is this comment accurate? If so, what is "the element" in this case?

File: apps/src/blockly/addons/workspaceSvgFrame.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  render() {
+    var minWidth = this.frameText_.getBoundingClientRect().width;
```
Comment (molly-moen) [inline apps/src/blockly/addons/workspaceSvgFrame.js]: nit: use `let` rather than `var`?

**Review by molly-moen — APPROVED** [review]
This looks good to me! Just a couple nits/clarifying questions

File: apps/src/blockly/addons/workspaceSvgFrame.js (hunk 2 above)
Comment (mikeharv) [reply apps/src/blockly/addons/workspaceSvgFrame.js:46]: Yes, but maybe it's not clear. The element is the workspace. The frame fits inside it. For `blockSvgFrame`, the element is a block and the frame stretches around it,

**Pull request merged**